src/lyra.ts

```diff
@@ -514,45 +517,87 @@ export function search<S extends PropertiesSchema>(
 
   const tokens = lyra.tokenizer.tokenizerFn!(params.term, language, false, lyra.tokenizer);
   const indices = getIndices(lyra, params.properties);
-  const uniqueDocIds = new Set<string>();
   const { limit = 10, offset = 0, exact = false } = params;
   const results: RetrievedDoc<S>[] = Array.from({
     length: limit,
   });
 
   const timeStart = getNanosecondsTime();
+  // uniqueDocsIDs contains unique document IDs for all the tokens in all the indices.
+  const uniqueDocsIDs: Set<string> = new Set();
+  // indexMap is an object containing all the indexes considered for the current search,
+  // and an array of doc IDs for each token in all the indices.
+  //
+  // Give the search term "quick brown fox" on the "description" index,
+  // indexMap will look like this:
+  //
+  // {
+  //   description: {
+  //     quick: [doc1, doc2, doc3],
+  //     brown: [doc2, doc4],
+  //     fox:   [doc2]
+  //   }
+  // }
+  const indexMap: IndexMap = {};
+  // After we create the indexMap, we need to calculate the intersection
+  // between all the postings lists for each token.
+  // Given the example above, docsIntersection will look like this:
+  //
+  // {
+  //   description: [doc2]
+  // }
+  //
+  // as doc2 is the only document present in all the postings lists for the "description" index.
+  const docsIntersection: TokenMap = {};
+
+  for (const index of indices) {
+    const tokensMap: TokenMap = {};
+    for (const token of tokens) {
+      tokensMap[token] = [];
+    }
+    indexMap[index] = tokensMap;
+    docsIntersection[index] = [];
+  }
 
-  let i = 0;
-  let j = 0;
-
-  for (const term of tokens) {
-    for (const index of indices) {
+  // Now it's time to loop over all the indices and get the documents IDs for every single term
+  for (const index of indices) {
+    for (const term of tokens) {
       const documentIDs = getDocumentIDsFromSearch(lyra, { ...params, index, term, exact });
 
       for (const id of documentIDs) {
-        uniqueDocIds.add(id);
+        indexMap[index][term].push(id);
```

Comment:
I think you can use one of:

* `indexMap[index][term].push(...documentIDs)`
* `indexMap[index][term] = indexMap[index][term].concat(documentIDs)` 

to avoid the for-loop.